File: backend/src/db/migrate.ts
import { readFileSync } from "fs";

import type Database from "better-sqlite3";

export const INITIAL_SCHEMA_VERSION = "001_init";

const INITIAL_SCHEMA_SQL = readFileSync(
  new URL("./migrations/001_init.sql", import.meta.url),
  "utf8",
);

export function runMigrations(database: Database.Database): void {
  database.exec(INITIAL_SCHEMA_SQL);

  database
    .prepare(
      `
      INSERT OR IGNORE INTO schema_migrations (version, applied_at)
      VALUES (?, ?)
      `,
    )
    .run(INITIAL_SCHEMA_VERSION, Date.now());
}

File: backend/internal/db/migrate.go
package db

import (
	"database/sql"
	"embed"
	"time"
)

//go:embed migrations/001_init.sql
var migrationsFS embed.FS

const InitialSchemaVersion = "001_init"

// Migrate mirrors backend/src/db/migrate.ts.
func Migrate(database *sql.DB) error {
	sqlBytes, err := migrationsFS.ReadFile("migrations/001_init.sql")
	if err != nil {
		return err
	}

	if _, err := database.Exec(string(sqlBytes)); err != nil {
		return err
	}

	_, err = database.Exec(
		`INSERT OR IGNORE INTO schema_migrations (version, applied_at) VALUES (?, ?)`,
		InitialSchemaVersion, time.Now().UnixMilli(),
	)
	return err
}
